### ISSUE TITLE
OFX Ver 1.x.x does not format SGML correctly

### ISSUE THREAD

**Issue by fergbrain** [csingley/ofxtools#38]
As far as I can tell (based on testing and the spec at http://www.ofx.net/downloads.html) tags that don't have additional nested tags have no end tag. For example downloading an OFX profile _should_ generate the following:

```
OFXHEADER:100
DATA:OFXSGML
VERSION:102
SECURITY:NONE
ENCODING:USASCII
CHARSET:NONE
COMPRESSION:NONE
OLDFILEUID:NONE
NEWFILEUID:4c8980e0-9664-40a9-9de1-2fa763615089

<OFX>
    <SIGNONMSGSRQV1>
        <SONRQ>
            <DTCLIENT>20180615225707
            <USERID>anonymous00000000000000000000000
            <USERPASS>anonymous00000000000000000000000
            <LANGUAGE>ENG
            <FI>
                <ORG>BECU
                <FID>3670
            </FI>
            <APPID>QWIN
            <APPVER>2300
        </SONRQ>
    </SIGNONMSGSRQV1>
    <PROFMSGSRQV1>
        <PROFTRNRQ>
        <TRNUID>097eefe1-8f81-43a2-8b43-cae62a774802
        <PROFRQ>
            <CLIENTROUTING>NONE
            <DTPROFUP>19900101000000
        </PROFRQ>
        </PROFTRNRQ>
    </PROFMSGSRQV1>
</OFX>
```

However, it instead generates the following:
```
OFXHEADER:100
DATA:OFXSGML
VERSION:102
SECURITY:NONE
ENCODING:USASCII
CHARSET:NONE
COMPRESSION:NONE
OLDFILEUID:NONE
NEWFILEUID:4c8980e0-9664-40a9-9de1-2fa763615089

<OFX>
    <SIGNONMSGSRQV1>
        <SONRQ>
            <DTCLIENT>20180615225707</DTCLIENT>
            <USERID>anonymous00000000000000000000000</USERID>
            <USERPASS>anonymous00000000000000000000000</USERPASS>
            <LANGUAGE>ENG</LANGUAGE>
            <FI>
                <ORG>BECU</ORG>
                <FID>3670</FID>
            </FI>
            <APPID>QWIN</APPID>
            <APPVER>2300</APPVER>
        </SONRQ>
    </SIGNONMSGSRQV1>
    <PROFMSGSRQV1>
        <PROFTRNRQ>
            <TRNUID>097eefe1-8f81-43a2-8b43-cae62a774802</TRNUID>
            <PROFRQ>
                <CLIENTROUTING>NONE</CLIENTROUTING>
                <DTPROFUP>19900101000000</DTPROFUP>
            </PROFRQ>
        </PROFTRNRQ>
    </PROFMSGSRQV1>
</OFX>
```

You can generate this using the following command: `ofxget -n https://onlinebanking.becu.org/ofx/ofxprocessor.asp --fid 3670 --org BECU --appver 2400 -p --version 102`

**Comment by csingley** [comment]
God, are you forced to used OFXv102?

Have you got a reference backing up your assertion?  I'm not aware of any SGML dialect where end tags are actually forbidden in any context.  The standard language I've seen is just about making certain tags optional, as a convenience.

In the 1.6 spec they added a section explicitly stating that end tags are optional for all elements... probably because of shitty OFX server implementations like (I'm assuming) your CU's.

**Comment by fergbrain** [comment]
> God, are you forced to used OFXv102?

As far as I can tell, yes I am. Fun times.

>Have you got a reference backing up your assertion?

The two things that caught my eye in the spec is section 1.2.2 ("Request and Response Model") and section 2.5.1.1 ("Signon Request <SONRQ>") from Open Financial Exchange Specification 1.0.2. I've uploaded the section 2.5.1.1 to make it easier for your reference but you can also get it at http://www.ofx.net/downloads.html:

![2 5 1 1](https://user-images.githubusercontent.com/1291387/41493918-7f487176-7104-11e8-8d1b-50d59816b62d.png)

If you note, DTCLIENT, USERID, USERPASS, LANGUAGE, APPID, and APPVER all do _not_ have closing tags noted, whereas SONRQ and FI do. This matches the behavior seen.

Section 1.2.2 also says,
> The key rule of Open Financial Exchange syntax is that each tag is either an element or an aggregate. Data follows its element tag. An aggregate tag begins a compound tag sequence, which must end with a matching tag; for example, <AGGREGATE> ... </AGGREGATE>. 

The term "aggregate" matches the description used in section 2.5.1.1 and requires the ending tag. I will also admit that I've not been able to find a place in the spec that explicitly says that element tags must _not_ end with a matching tag.

It's entirely possible that BECU hasn't correctly implemented the spec, however I started by using mitmproxy with Banktivity to see how they were (successfully) connecting and they don't use end tags...which is how I figured out the issue (I made the same initial assumption you did)

![mitmproxy_screenshot](https://user-images.githubusercontent.com/1291387/41494068-5646039a-7106-11e8-8767-5cec5af3ec85.png)

I also used mitmproxy to inspect how ofxtools was actually communicating with BECU and to try to get it to match how Banktivity did it, which is when I ran into the end tags issue described above.

I think the easiest way to verify this is by running the following:

curl -v -X POST --data-binary @ofx.txt --header "Content-Type: application/x-ofx" --header "Accept:*/*, application/x-ofx" https://onlinebanking.becu.org/ofx/ofxprocessor.asp

Swap out _bad and _good for...bad and good results. The _only_ difference is the ending tags:

[ofx_bad.txt](https://github.com/csingley/ofxtools/files/2107745/ofx_bad.txt)
[ofx_good.txt](https://github.com/csingley/ofxtools/files/2107746/ofx_good.txt)

**Comment by csingley** [comment]
I wouldn't get too hung up on the formatting of the human-readable OFX spec.  They use the same formatting to define aggregates in the OFXv2 series specs, which (being compliant XML) explicitly require end tags for everything.  In any case it's pretty thin basis to conclude that OFX102 intended the radical mutation of actually FORBIDDING clarity of element closure, thereby spawning a sui-generis phylum within the SGML taxonomy.

Rather, I direct your attention to section 2.8 of version 2.2 of the OFX spec:

> OFX 1.0.2 required any parser to accept end tags but did not require clients or servers to send elements with end tags.

But I wouldn't get too hung up on the spec, either, really.  The reality is that OFX server implementation compliance with the spec is generally pretty bad - largely a result of development efforts following "code to Quicken" rather than "code to spec".  Whoever wrote the OFX server that your CU is using probably saw that all the samples in the distributed document didn't close element tags, coded their parser to that, validated it against Quicken (which generated OFXv1 with unclosed element tags), and called it a day.

The code that's bothering you is in `ofxtools.Client.download()` (line 239), which simply calls `xml.etree.ElementTree.tostring()`.  What you want to do is test the generated OFX header version, and for OFXv1 call a different function instead (one that omits closing tags for elements).  I'd probably accept a patch like that (helps some people, shouldn't break anything, and performance is a nonissue here) but I'm not in a big hurry to write one.  It's probably not that bad, but I've still got PTSD from writing a regex parser for OFXv1.

**Comment by csingley** [comment]
I was offline for the weekend, so didn't see your patch and meanwhile implemented a different solution to your problem, which I pushed as [3128c8d](https://github.com/csingley/ofxtools/commit/3128c8d8fdd33c0f789cedbf12560161cb778ee6).

It makes unclosed end-tags an optional feature for OFXv1, with a CLI switch to enable it.  It does however break pretty-printing for the time being.  No great matter, but another user has a bank that won't accept OFX requests unless the tags have newlines between them.  I weep for the man-hours poured out into the sand by the OFX spec authors.

There's probably a bug or two in my implementation here; I've checked it in as a special branch just for you to play with.  Let me know if you find what I overlooked, or better yet fix it.  If you're well satisfied I'll merge it to master; it's pretty low impact, as long as it's working correctly.

Thanks.
